### PR TITLE
feat: show completion hint on the first suggestion

### PR DESCRIPTION
**Pull request by yairco1990**
https://user-images.githubusercontent.com/10290661/182035131-fe5b5afc-ca7e-4696-bdbe-04e47ac4fd12.mp4

### CODE REVIEW

File: src/main/java/com/tabnine/Initializer.java [hunk 1]
```diff
@@ -43,6 +44,7 @@ private void initialize() {
       CapabilitiesService.getInstance().init();
       TabnineUpdater.pollUpdates();
       PluginInstaller.addStateListener(instanceOfUninstallListener());
+      UserState.init();
```
Comment (yonip23) [inline src/main/java/com/tabnine/Initializer.java:47]: weren't we using `StateRequest` anywhere else? how aobut migrating existing places to use this singleton state?

File: src/main/java/com/tabnine/balloon/GotItTooltipActions.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package com.tabnine.balloon;
+
+public interface GotItTooltipActions {
```
Comment (yonip23) [inline src/main/java/com/tabnine/balloon/GotItTooltipActions.java:3]: its a single action, so `GotItTooltipAction`

**Review by yonip23 — CHANGES_REQUESTED** [review]
Overall good job, the UI looks quite good as well 🙂 

Other than my specific comments, there's something that doesn't sit well with me regarding the lifecycle of ths "got it" ui element - 
It is being disposed and re-rendered on every document change.
I would have expect to see something that is being rendered once upon first hint, and disposed once when we decide to dispose it.

It'll solve 2 issues that I see with the current implementation:
1. The message is "moving around" with the user when they type
2. The message is "blinking" - disappears and re-appears on every key stroke.

I'd also like to see some tests 🙂

**Comment by yairco1990** [comment]
This is how it looks now:
https://user-images.githubusercontent.com/10290661/182035062-7fb764c9-8769-4bba-aa18-04ccff72bea7.mp4

File: src/main/java/com/tabnine/Initializer.java (hunk 1 above)
Comment (yairco1990) [reply src/main/java/com/tabnine/Initializer.java:47]: I'm not sure that a stale state fits for the other usages. for me it's ok as I'm using only the installation time.

File: src/main/java/com/tabnine/balloon/GotItTooltip.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import com.intellij.openapi.util.Disposer;
+import com.intellij.ui.JBColor;
+import com.intellij.ui.awt.RelativePoint;
+import com.intellij.util.ui.JBUI;
+import com.tabnine.binary.BinaryRequestFacade;
+import com.tabnine.binary.requests.notifications.shown.HintShownRequest;
+import javax.swing.*;
+import org.jetbrains.annotations.NotNull;
+
+public class GotItTooltip implements Disposable {
```
Comment (yonip23) [inline src/main/java/com/tabnine/balloon/GotItTooltip.java:24]: rename class to `SuggestionHintTooltipUI` or something

File: src/main/java/com/tabnine/balloon/CompletionHintTooltip.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.editor.Editor;
+import com.tabnine.selections.CompletionListener;
+import com.tabnine.selections.CompletionObserver;
+import com.tabnine.state.UserState;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class CompletionHintTooltip {
```
Comment (yonip23) [inline src/main/java/com/tabnine/balloon/CompletionHintTooltip.java:11]: rename to `SuggestionHintTooltip`

File: src/main/java/com/tabnine/balloon/CompletionHintTooltip.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+          "first-completion-hint",
+          "Your first completion from Tabnine",
+          "Press <b>Tab</b> to accept Tabnine's suggestion",
+          () -> UserState.getInstance().getCompletionHintState().setIsShown(true));
+
+  public static void handle(Editor editor) {
+    try {
+      UserState userState = UserState.getInstance();
+      if (userState.getCompletionHintState().isEligibleForCompletionHint()) {
+        if (!completionHintTooltip.isVisible()) {
```
Comment (yonip23) [inline src/main/java/com/tabnine/balloon/CompletionHintTooltip.java:23]: invert conditions to something like
```java
if (!userState.getCompletionHintState().isEligibleForCompletionHint() || 
    completionHintTooltip.isVisible()) {
    // "else" code here
    return;
}
// main code here
```

File: src/main/java/com/tabnine/balloon/CompletionHintTooltip.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+              .schedule(
+                  () -> {
+                    UserState.getInstance().getCompletionHintState().setIsShown(true);
+                    completionHintTooltip.dispose();
+                  },
+                  30,
+                  TimeUnit.SECONDS);
+        }
+      } else {
+        userState.getCompletionHintState().setIsShown(true);
```
Comment (yonip23) [inline src/main/java/com/tabnine/balloon/CompletionHintTooltip.java:44]: not sure its important, but why not checking if its already true before setting? `setIsShown` is saving a persistent state so I guess is quite expensive operation

File: src/main/java/com/tabnine/balloon/GotItTooltip.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    panel.add(Box.createVerticalStrut(15));
+    panel.add(gotItButton);
+
+    return panel;
+  }
+
+  private void showTooltip(Editor editor, Balloon tooltip) {
+    RelativePoint relativePoint = JBPopupFactory.getInstance().guessBestPopupLocation(editor);
+    relativePoint.getPoint().translate(-5, -editor.getLineHeight() / 2);
```
Comment (yonip23) [inline src/main/java/com/tabnine/balloon/GotItTooltip.java:92]: extract -5 and 2 to consts

File: src/main/java/com/tabnine/balloon/GotItTooltipAction.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package com.tabnine.balloon;
+
+public interface GotItTooltipAction {
+  void onGotItClicked();
+}
```
Comment (yonip23) [inline src/main/java/com/tabnine/balloon/GotItTooltipAction.java:5]: rename to `SuggestionHintTooltipAction`.
the function as well - `onSuggestionHintTooltipClicked`

File: src/main/java/com/tabnine/capabilities/Capability.java [hunk 9]
```diff
@@ -7,4 +7,7 @@ public enum Capability {
   INLINE_SUGGESTIONS,
   @SerializedName("alpha")
   ALPHA,
+
+  @SerializedName("completion_hint_enabled")
+  COMPLETION_HINT_ENABLED,
```
Comment (yonip23) [inline src/main/java/com/tabnine/capabilities/Capability.java:12]: `SUGGESTION_HINT_ENABLED` - also the value in the binary should be changed

File: src/main/java/com/tabnine/general/Utils.java [hunk 10]
```diff
@@ -70,4 +72,21 @@ public static List<String> asLines(String block) {
...
+
+  public static String wrapWithHtmlTag(String content, String tag) {
+    return "<" + tag + ">" + content + "</" + tag + ">";
+  }
+
+  public static long getDaysDiff(Date date1, Date date2) {
+    try {
+      return TimeUnit.DAYS.convert(
+          Math.abs(date2.getTime() - date1.getTime()), TimeUnit.MILLISECONDS);
+    } catch (NullPointerException e) {
```
Comment (yonip23) [inline src/main/java/com/tabnine/general/Utils.java:88]: why catching nulls here? you can annotate the dates with `@NotNull`

File: src/main/java/com/tabnine/balloon/GotItTooltip.java (hunk 3 above)
Comment (yairco1990) [reply src/main/java/com/tabnine/balloon/GotItTooltip.java:24]: the purpose of this class is to provide other hints we may use in the future, and not just the completion hint

File: src/main/java/com/tabnine/balloon/CompletionHintTooltip.java (hunk 5 above)
Comment (yairco1990) [reply src/main/java/com/tabnine/balloon/CompletionHintTooltip.java:23]: in this specific example, I want to handle the else case of the isEligible condition, and ignore the else case of the isVisible

Comment (yairco1990) [reply src/main/java/com/tabnine/balloon/CompletionHintTooltip.java:23]: something like
`      if (!userState.getCompletionHintState().isEligibleForCompletionHint()) {
        userState.getCompletionHintState().setIsShown(true);
        return;
      }
      if (!completionHintTooltip.isVisible()) {
        return;
      }`

File: src/main/java/com/tabnine/balloon/CompletionHintTooltip.java (hunk 6 above)
Comment (yairco1990) [reply src/main/java/com/tabnine/balloon/CompletionHintTooltip.java:44]: you're right. I'll add the checking

File: src/main/java/com/tabnine/capabilities/Capability.java [hunk 11]
```diff
@@ -7,4 +7,7 @@ public enum Capability {
   INLINE_SUGGESTIONS,
   @SerializedName("alpha")
   ALPHA,
+
+  @SerializedName("suggestion_hint_enabled")
```
Comment (yonip23) [inline src/main/java/com/tabnine/capabilities/Capability.java:11]: first_suggestion_hint?

**Review by yonip23 — APPROVED** [review]
🚀

File: src/main/java/com/tabnine/capabilities/Capability.java (hunk 11 above)
Comment (yonip23) [reply src/main/java/com/tabnine/capabilities/Capability.java:11]: im worried that we'll have a feature which includes suggestion hints, but unrelated to this feature at all - then we'll have a bit of a hard time with the naming convention. 
i'll leave it to your decision, not critical 🙂

Comment (yairco1990) [reply src/main/java/com/tabnine/capabilities/Capability.java:11]: yeah, that could happen. yalla, I'm changing it. :)

**Pull request merged**